### PR TITLE
Add functionality to read attributes from content apps

### PR DESCRIPTION
**Pull request by amitnj**
#### Problem
Functionality for reading attributes from content apps was missing in the tv-app example

#### Change overview
Added mechanism to send attribute read requests to content app and send the response back to the client. One cluster attributes were implemented as an example.

#### Testing
Manually using the tv-casting-app

### CODE REVIEW

**Review by chrisdecenzo — APPROVED** [review]

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        handlerContext.SetCommandHandled();
+        break;
+    }
+
+    // case app::Clusters::TargetNavigator::Id:
+    //     break;
+
+    // case app::Clusters::MediaPlayback::Id:
+    //     break;
+
```
Comment (lazarkov) [inline examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:129]: can be removed?

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            ChipLogError(Zcl, "Java exception in ContentAppAttributeDelegate::Read");
+            env->ExceptionDescribe();
+            env->ExceptionClear();
+            //            FormatResponseData(handlerContext, "{\"value\":{}}");
+            return "";
+        }
+        const char * respStr = env->GetStringUTFChars(resp, 0);
+        ChipLogProgress(Zcl, "ContentAppAttributeDelegate::Read got response %s", respStr);
+        //        FormatResponseData(handlerContext, respStr);
```
Comment (lazarkov) [inline examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:61]: can be removed?

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        jstring resp =
+            (jstring) env->CallObjectMethod(mContentAppEndpointManager, mReadAttributeMethod, static_cast<jint>(aPath.mEndpointId),
+                                            static_cast<jint>(aPath.mClusterId), static_cast<jint>(aPath.mAttributeId));
+        if (env->ExceptionCheck())
+        {
+            ChipLogError(Zcl, "Java exception in ContentAppAttributeDelegate::Read");
+            env->ExceptionDescribe();
+            env->ExceptionClear();
+            //            FormatResponseData(handlerContext, "{\"value\":{}}");
```
Comment (lazarkov) [inline examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:56]: can be removed?

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp (hunk 1 above)
Comment (amitnj) [reply examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:129]: Added as marker for which other responses have to be de-serialized.

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp (hunk 2 above)
Comment (amitnj) [reply examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:61]: Will remove

Comment (amitnj) [reply examples/tv-app/android/java/ContentAppAttributeDelegate.cpp]: removed.

**Review by lazarkov — APPROVED** [review]

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
```
[the remark this answers is not in the document]
Comment (amitnj) [reply examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:3]: I can change this across the board in a separate PR.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp [hunk 5]
```diff
@@ -73,18 +74,38 @@ void AppContentLauncherManager::HandleLaunchUrl(CommandResponseHelper<LaunchResp
...
+                                               chip::app::Clusters::ContentLauncher::Attributes::AcceptHeader::Id);
+    const char * resStr = mAttributeDelegate->Read(aPath);
+    ChipLogProgress(Zcl, "AppContentLauncherManager::HandleGetSupportedStreamingProtocols response %s", resStr);
+
+    if (resStr != nullptr && *resStr != 0)
+    {
+        Json::Reader reader;
+        Json::Value value;
+        reader.parse(resStr, value);
+        const char * attrId = to_string(chip::app::Clusters::ContentLauncher::Attributes::AcceptHeader::Id).c_str();
```
Comment (andy31415) [inline examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp:96]: what is the lifetime for to_string? 

It seems to me that we create a temporary with to_string, grabs its c_str ... but after that the temporary is deleted. Is that true? if so you need to store the std::string to ensure c_str is still valid.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp [hunk 6]
```diff
@@ -98,5 +119,26 @@ CHIP_ERROR AppContentLauncherManager::HandleGetAcceptHeaderList(AttributeValueEn
...
+
+    if (resStr == nullptr || *resStr == 0)
+    {
+        return mSupportedStreamingProtocols;
+    }
+
+    Json::Reader reader;
+    Json::Value value;
+    reader.parse(resStr, value);
+    const char * attrId = to_string(chip::app::Clusters::ContentLauncher::Attributes::SupportedStreamingProtocols::Id).c_str();
```
Comment (andy31415) [inline examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp:135]: same comment here: what ensures c_str lifetime stays valid?

File: examples/tv-app/android/java/AppImpl.h [hunk 7]
```diff
@@ -86,9 +86,10 @@ class DLL_EXPORT ContentAppImpl : public ContentApp
                    const char * szApplicationVersion, const char * setupPIN, jobject manager) :
         mApplicationBasicDelegate(kCatalogVendorId, BuildAppId(vendorId), szVendorName, vendorId, szApplicationName, productId,
                                   szApplicationVersion),
-        mAccountLoginDelegate(setupPIN), mContentLauncherDelegate(ContentAppCommandDelegate(manager), { "image/*", "video/*" },
-                                                                  to_underlying(SupportedStreamingProtocol::kDash) |
-                                                                      to_underlying(SupportedStreamingProtocol::kHls)),
+        mAccountLoginDelegate(setupPIN),
+        mContentLauncherDelegate(new ContentAppAttributeDelegate(manager), { "image/*", "video/*" },
```
Comment (andy31415) [inline examples/tv-app/android/java/AppImpl.h:90]: Where is the object that is allocated here via `new` freed?

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#include <zap-generated/endpoint_config.h>
+
+namespace chip {
+namespace AppPlatform {
+
+using LaunchResponseType = chip::app::Clusters::ContentLauncher::Commands::LaunchResponse::Type;
+
+const char * ContentAppAttributeDelegate::Read(const chip::app::ConcreteReadAttributePath & aPath)
+{
+    if (aPath.mEndpointId >= FIXED_ENDPOINT_COUNT)
```
Comment (andy31415) [inline examples/tv-app/android/java/ContentAppAttributeDelegate.cpp:39]: if you switch the order of conditions, indent is reduced and code is easier to read. Do this:

```
if (aPath.mEndpointId < FIXED_ENDPOINT_COUNT) {
    return "";
}

JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
// ....
```

File: examples/tv-app/android/java/AppImpl.h (hunk 7 above)
Comment (amitnj) [reply examples/tv-app/android/java/AppImpl.h]: Dont think I need to create object here using new. Removed.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp (hunk 6 above)
Comment (amitnj) [reply examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp]: Created a local variable.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp (hunk 5 above)
Comment (amitnj) [reply examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp]: Created a local variable.

File: examples/tv-app/android/java/ContentAppAttributeDelegate.cpp (hunk 8 above)
Comment (amitnj) [reply examples/tv-app/android/java/ContentAppAttributeDelegate.cpp]: Updated.

File: examples/tv-app/android/java/ContentAppCommandDelegate.cpp [hunk 9]
```diff
@@ -110,9 +111,8 @@ void ContentAppCommandDelegate::InvokeCommand(CommandHandlerInterface::HandlerCo
 void ContentAppCommandDelegate::FormatResponseData(CommandHandlerInterface::HandlerContext & handlerContext, const char * response)
 {
     Json::Reader reader;
-    Json::Value resJson;
-    reader.parse(response, resJson);
-    Json::Value value = resJson["value"];
+    Json::Value value;
+    reader.parse(response, value);
```
Comment (andy31415) [inline examples/tv-app/android/java/ContentAppCommandDelegate.cpp:115]: parse returns a bool for success/failure. Please handle failure.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp [hunk 10]
```diff
@@ -98,6 +120,27 @@ CHIP_ERROR AppContentLauncherManager::HandleGetAcceptHeaderList(AttributeValueEn
...
+    ChipLogProgress(Zcl, "AppContentLauncherManager::HandleGetSupportedStreamingProtocols response %s", resStr);
+
+    if (resStr == nullptr || *resStr == 0)
+    {
+        return mSupportedStreamingProtocols;
+    }
+
+    Json::Reader reader;
+    Json::Value value;
+    reader.parse(resStr, value);
```
Comment (andy31415) [inline examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp:135]: handle parse failures

**Review by woody-apple — APPROVED** [review]

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp [hunk 11]
```diff
@@ -73,18 +74,39 @@ void AppContentLauncherManager::HandleLaunchUrl(CommandResponseHelper<LaunchResp
...
+                                               chip::app::Clusters::ContentLauncher::Attributes::AcceptHeader::Id);
+    const char * resStr = mAttributeDelegate.Read(aPath);
+    ChipLogProgress(Zcl, "AppContentLauncherManager::HandleGetSupportedStreamingProtocols response %s", resStr);
+
+    if (resStr != nullptr && *resStr != 0)
+    {
+        Json::Reader reader;
+        Json::Value value;
+        reader.parse(resStr, value);
+        std::string attrId = to_string(chip::app::Clusters::ContentLauncher::Attributes::AcceptHeader::Id).c_str();
```
[the remark this answers is not in the document]
Comment (amitnj) [reply examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp]: That was the plan. I did it in one place and missed here. Fixed.

File: examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp [hunk 12]
```diff
@@ -73,18 +74,39 @@ void AppContentLauncherManager::HandleLaunchUrl(CommandResponseHelper<LaunchResp
...
+    chip::app::ConcreteReadAttributePath aPath(mEndpointId, chip::app::Clusters::ContentLauncher::Id,
+                                               chip::app::Clusters::ContentLauncher::Attributes::AcceptHeader::Id);
+    const char * resStr = mAttributeDelegate.Read(aPath);
+    ChipLogProgress(Zcl, "AppContentLauncherManager::HandleGetSupportedStreamingProtocols response %s", resStr);
+
+    if (resStr != nullptr && *resStr != 0)
+    {
+        Json::Reader reader;
+        Json::Value value;
+        reader.parse(resStr, value);
```
[the remark this answers is not in the document]
Comment (amitnj) [reply examples/tv-app/android/include/content-launcher/AppContentLauncherManager.cpp]: Added failure in parsing by sending the default attribute value.

File: examples/tv-app/android/java/ContentAppCommandDelegate.cpp (hunk 9 above)
Comment (amitnj) [reply examples/tv-app/android/java/ContentAppCommandDelegate.cpp]: Added failure handling for parsing.

**Pull request merged**